design/rich-authorization-requests.md

```diff
@@ -0,0 +1,173 @@
+# Rich Authorization Requests (RAR)
+
+* **Status**: Notes
+* **JIRA**: TBD
+* **Implementation**: https://github.com/tnorimat/keycloak/pull/24 
+
+
+## Motivation
+
+[Rich Authorization Requests][1] is designed to allow OAuth Client to include detailed information to an authorization request. Thus, Client can make a specific authorization requirement.
+
+## Specifications
+
+The spec introduces a new parameter "authorization_details" that allows clients to specify their fine-grained authorization, such as `please let me make a payment with the amount of 45 Euros` or `please give me read access to folder A and write access to file X`.
+
+An example of authorization details using
+the payment authorization object:
+
+````
+[
+      {
+         "type": "payment_initiation",
+         "actions": [
+            "initiate",
+            "status",
+            "cancel"
+         ],
+         "locations": [
+            "https://example.com/payments"
+         ],
+         "instructedAmount": {
+            "currency": "EUR",
+            "amount": "123.50"
+         },
+         "creditorName": "Merchant123",
+         "creditorAccount": {
+            "iban": "DE02100100109307118603"
+         },
+         "remittanceInformationUnstructured": "Ref Number Merchant"
+      }
+   ]
+````
+
+The type of data contained in authorization details is infinite and depends on the field (Payment, Health, Tax Data, etc...)
+
+Therefore, it will be more scalable to add the processing of authorization details as SPI (e.g. `RarProcessor.Class`). Then, an extension will come with the specific implementation.
+
+However, there is some default configuration and behavior which should be implemented:
+
+- Add parameters in Authorization Server Metadata
+- Add parameters in Client Metadata
+- Accept `authorization_details` parameter in all places where the `scope` parameter is used
+- Add authorization details to access tokens and token introspection
+
+### Authorization Server Metadata
+
+This parameter must be added to the .well-known/openid-configuration:
+
+- **authorization_details_types_supported**
+
+The authorization data types supported can be determined using the metadata parameter "authorization_details_types_supported", which is an JSON array.
```

Comment:
In my opinion, it's way around, a client should pick the types it wants to use in `authorization_details_types_supported`.  published in AS metadata

ex:

AS `authorization_details_types_supported: https://www.someorg.com/account_information, https://www.someorg.com/payment_initiation`

a client can choose to use just account_information ie `authorization_details_type: https://www.someorg.com/account_information`

Maybe we should also check in the dynamic client registration if `authorization_details_type` are compliant with `authorization_details_types_supported`. because it should normally.

Reply:
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-rar#section-10

Reply:
+1 to the point that discovery claim `authorization_details_types_supported` will be an aggregated list of all types across all types of processor providers. Regarding collisions, I supposed we can add something like `getPriority` to the processor methods? This will allow that custom processor provider will be chosen for processing when dealing with the particular conflicting type.

For example: Assume that there will be built-in provider `DefaultProcessor`, which will support types `my-type-1` and `my-type-2` with priority 1. Then assume that administrator will deploy his own custom provider with higher priority (EG. `MyCustomProcessor`), which will support types `my-type-2` and `my-type-3`  and will have higher priority.

This will ensure that:
- The type `my-type-1` will be processed by default processor
- The type `my-type-3` will be processed by MyCustomProcessor
- The type `my-type-2` exists in both processors, but since `MyCustomProcessor` has higher priority, it will be chosen for processing this request.

Reply:
It seems to me that priority of the processor can be either hardcoded in the implementation or overriden in the standalone(-*).xml . Which means configurable for the whole server. It seems to me that have the processor priority specific for the whole server is fine for most of the use-cases and we don't need the realm or client level.

@thomasdarimont @guymoyo WDYT? Do you have any use-case for have the ability to choose the processor for the realm or client level?

Reply:
Hi @mposolda, thanks for the explanation, it's clear to me. I don't have a particular use case at the realm or client level.
I will add your text as an explanation for `authorization_details_types_supported`.